packages/virtual/src/index.tsx

```diff
@@ -1,71 +1,123 @@
-  class?: string;
-  each: T | undefined | null | false;
-  overscanCount?: number;
-  rootHeight: number;
-  rowHeight: number;
-}): JSX.Element {
-  let rootElement!: HTMLDivElement;
+export function createVirtualList<T extends readonly any[]>({
+  items,
+  rootHeight,
+  rowHeight,
+  overscanCount,
+}: VirtualListConfig<T>): VirtualListReturn<T> {
+  items = access(items) || ([] as any as T);
+  rootHeight = access(rootHeight);
+  rowHeight = access(rowHeight);
+  overscanCount = access(overscanCount) || 1;
 
   const [offset, setOffset] = createSignal(0);
-  const items = () => props.each || ([] as any as T);
 
-  const getFirstIdx = () =>
-    Math.max(0, Math.floor(offset() / props.rowHeight) - (props.overscanCount || 1));
+  const getFirstIdx = () => Math.max(0, Math.floor(offset() / rowHeight) - overscanCount);
 
   const getLastIdx = () =>
     Math.min(
-      items().length,
-      Math.floor(offset() / props.rowHeight) +
-        Math.ceil(props.rootHeight / props.rowHeight) +
-        (props.overscanCount || 1),
+      items.length,
+      Math.floor(offset() / rowHeight) + Math.ceil(rootHeight / rowHeight) + overscanCount,
     );
 
+  return [
+    {
+      containerHeight: () => items.length * rowHeight,
+      viewerTop: () => getFirstIdx() * rowHeight,
+      visibleItems: () => items.slice(getFirstIdx(), getLastIdx()) as unknown as T,
+    },
+    e => {
+      setOffset(e.target.scrollTop);
+    },
+  ];
+}
+
+type VirtualListProps<T extends readonly any[], U extends JSX.Element> = {
+  children: (item: T[number], index: Accessor<number>) => U;
+  each: T | undefined | null | false;
+  fallback?: JSX.Element;
+  overscanCount?: number;
+  rootHeight: number;
+  rowHeight: number;
+};
+
+/**
+ * A basic, unstyled virtualized list (see https://www.patterns.dev/vanilla/virtual-lists/) component you can drop into projects without modification
+ *
+ * @param children the flowComponent that will be used to transform the items into rows in the list
+ * @param each the list of items
+ * @param fallback the optional fallback to display if the list of items to display is empty
+ * @param overscanCount the number of elements to render both before and after the visible section of the list, so passing 5 will render 5 items before the list, and 5 items after. Defaults to 1, cannot be set to zero. This is necessary to hide the blank space around list items when scrolling
+ * @param rootHeight the height of the root element of the virtualizedList itself
+ * @param rowHeight the height of individual rows in the virtualizedList
+ * @returns virtualized list component
+ */
+export function VirtualList<T extends readonly any[], U extends JSX.Element>(
+  props: VirtualListProps<T, U>,
+): JSX.Element {
+  const [{ containerHeight, viewerTop, visibleItems }, onScroll] = createVirtualList({
+    items: props.each,
+    rootHeight: props.rootHeight,
+    rowHeight: props.rowHeight,
+    overscanCount: props.overscanCount,
+  });
```

Comment:
```suggestion
  const [{ containerHeight, viewerTop, visibleItems }, onScroll] = createVirtualList({
    items: () => props.each,
    rootHeight: () => props.rootHeight,
    rowHeight: () => props.rowHeight,
    overscanCount: () => props.overscanCount,
  });
```

Also it's a shame that you cannot just pass `props` directly, simply because `items` has a different name. Not sure what to do about it though. Maybe `createVirtualList` could accept both and figure out which one was picked. You'd need to not destructure `props` in `createVirtualList` for that though.

Reply:
That is annoying, but that API detail won't be visible to consumers, so I don't want to add any logic that complicates `createVirtualList` to avoid just having to rename a var in `VirtualList`.

Why is it important that arguments be changed to `() =>`s here? `createVirtualList` takes `MaybeAccessor`s and uses `access` on them (your suggestion) which I think should handle them being passed as simple arguments here. The tests all work too.